libmusig2.c

```diff
@@ -1,282 +1,359 @@
+        {
+            printf("Failed to aggregate signature!\n");
+            return;
         }
 
-        crypto_core_ristretto255_add(aggr_pk, aggr_pk, temp_point);
+        /* Copy the content of aggRound1->OUT_R_LIST into aggRound1->OUT.
+         * This is done to generate a hash of output of first round signature.
+         * See SET_Param(). */
+        for(i=0;i<PK_SIZE;i++,k++){
+            aggRound1->OUT[k] = aggRound1->OUT_R_LIST[j][i];
+        }
     }
-
-    return 0;
 }
 
-// Given as input a MuSig2 aggregated key, a secret key, a (unused) secret nonce, the commitments of other parties
-// secret nonces, the signers public exponent, and a message, return the partial signature corresponding to the given
-// message.
-void partial_signature(unsigned char *sig,
-                      unsigned char *aggr_announcement,
-                      const unsigned char *aggr_pk,
-                      const unsigned char *sk,
-                      const unsigned char *secret_nonce,
-                      const unsigned char *committed_nonces,
-                      const unsigned char *own_exponent,
-                      const unsigned char *m,
-                      unsigned long long mlen,
-                      unsigned long nr_signers) {
-    // First, we combine the committed_nonces from all participants, by adding all of them.
-    unsigned char aggr_comms[NR_V * crypto_core_ristretto255_BYTES] = {0};
-    for (int j = 0; j < NR_V; j++) {
-        for (int i = 0; i < nr_signers; i++) {
-            crypto_core_ristretto255_add(
-                    aggr_comms + (j * crypto_core_ristretto255_BYTES),
-                    aggr_comms + (j * crypto_core_ristretto255_BYTES),
-                    committed_nonces + i * NR_V * crypto_core_ristretto255_BYTES + j * crypto_core_ristretto255_BYTES
-            );
+void AGG_Key(const secp256k1_context* ctx, AGG_ROUND1 aggRound1, PARAM_ROUND2 paramRound2){
+
+    int i;
+    size_t point_ser_len = PK_SIZE;
+    unsigned char a[32];
+
+    secp256k1_pubkey** temp_Xi = (secp256k1_pubkey**)malloc(sizeof(secp256k1_pubkey*)*N);
+    secp256k1_pubkey* temp_X_ = (secp256k1_pubkey*)malloc(sizeof(secp256k1_pubkey));
+
+    for(i=0;i<N;i++){
+        temp_Xi[i] = (secp256k1_pubkey*)malloc(sizeof(secp256k1_pubkey));
+
+        /* Parse the public key in secp256k1_pubkey object.
+         * Compute the key aggregate coefficient (a) of signer i.
+         * Multiply the public key of signer i with a, store it in secp256k1_pubkey array. */
+        if (!(secp256k1_ec_pubkey_parse(ctx, temp_Xi[i], aggRound1->PK_LIST[i], point_ser_len) &&
+                secp256k1_tagged_sha256(ctx, a, aggRound1->PK_LIST[i], PK_SIZE, paramRound2->L, PK_SIZE*N) &&
+                secp256k1_ec_pubkey_tweak_mul(ctx, temp_Xi[i], a)))
+        {
+            printf("Failed to generate key aggregation coefficients!\n");
```

Comment:
when does this happen? what happens if a > p? mod reduction? Careful with the uniformity of the reduced scalar (see this [section](https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-10.html#name-security-considerations-3)). Also, consider domain separation for the distinct hashes.